### ISSUE TITLE
train.txt in train.py

### ISSUE THREAD

**Issue by james20141606** [r9y9/wavenet_vocoder#165]
Hey, great work on vocoder! I tried your repo to train a model by myself and I found a mistake in https://github.com/r9y9/wavenet_vocoder/blob/cd2abdfd221c350e9d94184a454213ec55eae440/train.py
If I did not generate a train.txt using preprocess.py. I think I will go through the if condition in:
```
def collect_files(self):
        meta = join(self.dump_root, "train.txt")
        if not exists(meta):
            paths = sorted(glob(join(self.dump_root, "*-{}.npy".format(self.typ))))
            return paths
```
however, it still needs meta in the following:
```
with open(meta, "rb") as f:
            lines = f.readlines()
```
I am not sure if you should add an else condition here to skip this part if I do not need train.txt

**Comment by r9y9** [comment]
I'm not sure if I understand you correctly. train.txt should be generated if you use preprocess.py. As you can see in the first snippet you put in the description, there's a return expression right after the `if not` condition, so `else` branch shouldn't be needed. What are you trying to do?

**Comment by james20141606** [comment]
Thanks for the reply! I mean if I did not use preprocess.py but use my own preprocess programs, I won't have any train.txt file. Then in collect_files function, should the program only execute the if not part? However in your program, if I didnot have the train.txt part, I think the program will continue to run including `with open(meta, "rb") as f:` which will generate an error.

**Comment by r9y9** [comment]
If you want to use your own preprocess script, please do the same as preprocess.py and create the train.txt file. Or train.py doesn't work at all. 

The case we don't have train.txt is for just *inference* time. e.g., using tacotron or similar to use generate mel-spectrograms, and feed them to the wavenet vocoder to generate waveforms.

**Comment by james20141606** [comment]
Thank you for your reply! I fixed that problem and I have another problem: I found you hard code two phases for train data: "train_no_dev", "dev". Could you explain the difference between the data? Thank you very much!

**Comment by r9y9** [comment]
- train_no_dev: training set
- dev: validation set

**Comment by james20141606** [comment]
Thanks for the clarification! So should I prepare two train.txt including their corresponding filenames in each dir for training and validation? Or should I prepare only one train.txt file including all the filenames like this?
```
data/
├── dev
├── train_no_dev
└── train.txt
```

**Comment by r9y9** [comment]
You need to create train.txt for each directory. Please refer the one of examples in the repo https://github.com/r9y9/wavenet_vocoder/tree/master/egs.

```
$ tree dump/lj_mol/logmelspectrogram/norm -d 
dump/lj_mol/logmelspectrogram/norm
├── dev
├── eval
└── train_no_dev
```

https://github.com/r9y9/wavenet_vocoder#getting-started

**Comment by james20141606** [comment]
thanks again!
I have another problem about assert_ready_for_upsampling in collate_fn(batch) function. In your case if seems you should ensure len(c) (duration of spectrogram?) * hop size = len(x) (sample number of audio?). I use my own preprocess method and the len(x)*hop size and len(x) are not exactly same, they have a slight difference. Is it possible I skip this assertion or do some other operation?

**Comment by r9y9** [comment]
You should ensure that len(x)*hop size equals to len(x).
